File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/Fast/FastStringOutputTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if FAST_FMT
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.FormatterTests.Fast {
	public sealed class FastStringOutputTests {
		[Theory]
		[InlineData(-1)]
		[InlineData(0)]
		[InlineData(1)]
		[InlineData(1000)]
		void AppendChars(int capacity) {
			var output = capacity < 0 ? new FastStringOutput() : new FastStringOutput(capacity);
			Assert.Empty(output.ToString());
			output.Append('a');
			Assert.Equal("a", output.ToString());
			output.Append('b');
			Assert.Equal("ab", output.ToString());
			for (int i = 0; i < 1000; i++)
				output.Append('c');
			Assert.Equal("ab" + new string('c', 1000), output.ToString());
		}

		[Theory]
		[InlineData("")]
		[InlineData("q")]
		[InlineData("qw")]
		[InlineData("qwerty")]
		void AppendString(string s) {
			var capacities = new int[] { -1, 0, 1, 1000 };
			foreach (var capacity in capacities) {
				var output = capacity < 0 ? new FastStringOutput() : new FastStringOutput(capacity);
				Assert.Empty(output.ToString());
				output.Append(s);
				Assert.Equal(s, output.ToString());
				output.Append("abc");
				Assert.Equal(s + "abc", output.ToString());
				for (int i = 0; i < 200; i++)
					output.Append("x");
				Assert.Equal(s + "abc" + new string('x', 200), output.ToString());
				for (int i = 0; i < 200; i++)
					output.Append("yy");
				Assert.Equal(s + "abc" + new string('x', 200) + new string('y', 400), output.ToString());
			}
		}

		[Fact]
		void AppendNullWorks() {
			var output = new FastStringOutput();
			Assert.Empty(output.ToString());
			output.Append(null);
			Assert.Empty(output.ToString());
			output.Append('a');
			Assert.Equal("a", output.ToString());
			output.Append(null);
			Assert.Equal("a", output.ToString());
		}

		[Fact]
		void ClearWorks() {
			var output = new FastStringOutput();
			Assert.Empty(output.ToString());
			output.Append('a');
			Assert.Equal("a", output.ToString());
			output.Clear();
			Assert.Empty(output.ToString());
			output.Append("abc");
			Assert.Equal("abc", output.ToString());
			output.Clear();
			Assert.Empty(output.ToString());
		}

		[Fact]
		void LengthWorks() {
			var output = new FastStringOutput();
			Assert.Equal(0, output.Length);
			output.Append('a');
			Assert.Equal(1, output.Length);
			output.Append('b');
			Assert.Equal(2, output.Length);
			output.Append("cde");
			Assert.Equal(5, output.Length);
			output.Clear();
			Assert.Equal(0, output.Length);
			output.Append("abc");
			Assert.Equal(3, output.Length);
			output.Clear();
			Assert.Equal(0, output.Length);
		}

#if HAS_SPAN
		[Fact]
		void AsSpanWorks() {
			var output = new FastStringOutput();
			Assert.Empty(output.AsSpan().ToString());
			output.Append('a');
			Assert.Equal("a", output.AsSpan().ToString());
			output.Append('b');
			Assert.Equal("ab", output.AsSpan().ToString());
			for (int i = 0; i < 1000; i++)
				output.Append('c');
			Assert.Equal("ab" + new string('c', 1000), output.AsSpan().ToString());
			output.Clear();
			Assert.Empty(output.AsSpan().ToString());
		}
#endif

		[Theory]
		[InlineData(0, "", "abcdefgh")]
		[InlineData(8, "", "abcdefgh")]
		[InlineData(0, "x", "xbcdefgh")]
		[InlineData(3, "x", "abcxefgh")]
		[InlineData(7, "x", "abcdefgx")]
		[InlineData(0, "xyz", "xyzdefgh")]
		[InlineData(3, "xyz", "abcxyzgh")]
		[InlineData(5, "xyz", "abcdexyz")]
		void CopyToWorks(int index, string data, string expected) {
			var charData = "abcdefgh".ToCharArray();
			var output = new FastStringOutput();
			output.Append(data);
			output.CopyTo(charData, index);
			var s = new string(charData);
			Assert.Equal(expected, s);
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/FastStringOutputTests.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

import com.github.icedland.iced.x86.fmt.fast.FastStringOutput;

final class FastStringOutputTests {
	private static String createString(char c, int count) {
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < count; i++)
			sb.append(c);
		return sb.toString();
	}

	@ParameterizedTest
	@ValueSource(ints = { -1, 0, 1, 1000 })
	void appendChars(int capacity) {
		FastStringOutput output = capacity < 0 ? new FastStringOutput() : new FastStringOutput(capacity);
		assertEquals("", output.toString());
		output.append('a');
		assertEquals("a", output.toString());
		output.append('b');
		assertEquals("ab", output.toString());
		for (int i = 0; i < 1000; i++)
			output.append('c');
		assertEquals("ab" + createString('c', 1000), output.toString());
	}

	@ParameterizedTest
	@ValueSource(strings = { "", "q", "qw", "qwerty" })
	void appendString(String s) {
		int[] capacities = new int[] { -1, 0, 1, 1000 };
		for (int capacity : capacities) {
			FastStringOutput output = capacity < 0 ? new FastStringOutput() : new FastStringOutput(capacity);
			assertEquals("", output.toString());
			output.append(s);
			assertEquals(s, output.toString());
			output.append("abc");
			assertEquals(s + "abc", output.toString());
			for (int i = 0; i < 200; i++)
				output.append("x");
			assertEquals(s + "abc" + createString('x', 200), output.toString());
			for (int i = 0; i < 200; i++)
				output.append("yy");
			assertEquals(s + "abc" + createString('x', 200) + createString('y', 400), output.toString());
		}
	}

	@Test
	void appendNullWorks() {
		FastStringOutput output = new FastStringOutput();
		assertEquals("", output.toString());
		output.append(null);
		assertEquals("", output.toString());
		output.append('a');
		assertEquals("a", output.toString());
		output.append(null);
		assertEquals("a", output.toString());
	}

	@Test
	void clearWorks() {
		FastStringOutput output = new FastStringOutput();
		assertEquals("", output.toString());
		output.append('a');
		assertEquals("a", output.toString());
		output.clear();
		assertEquals("", output.toString());
		output.append("abc");
		assertEquals("abc", output.toString());
		output.clear();
		assertEquals("", output.toString());
	}

	@Test
	void lengthWorks() {
		FastStringOutput output = new FastStringOutput();
		assertEquals(0, output.size());
		output.append('a');
		assertEquals(1, output.size());
		output.append('b');
		assertEquals(2, output.size());
		output.append("cde");
		assertEquals(5, output.size());
		output.clear();
		assertEquals(0, output.size());
		output.append("abc");
		assertEquals(3, output.size());
		output.clear();
		assertEquals(0, output.size());
	}
}
